providers/rpc/rpc.go

```diff
@@ -0,0 +1,369 @@
+		return false, fmt.Errorf("error from rpc consumer: %v", resp.Error)
+	}
+
+	return true, nil
+}
+
+// PowerSet sets the power state of a BMC machine.
+func (p *Provider) PowerSet(ctx context.Context, state string) (ok bool, err error) {
+	switch strings.ToLower(state) {
+	case "on", "off", "cycle":
+		rp := RequestPayload{
+			ID:     time.Now().UnixNano(),
+			Host:   p.Host,
+			Method: PowerSetMethod,
+			Params: PowerSetParams{
+				State: strings.ToLower(state),
+			},
+		}
+		resp, err := p.process(ctx, rp)
+		if err != nil {
+			return ok, err
+		}
+		if resp.Error != nil && resp.Error.Code != 0 {
+			return ok, fmt.Errorf("error from rpc consumer: %v", resp.Error)
+		}
+
+		return true, nil
+	}
+
+	return false, errors.New("requested power state is not supported")
+}
+
+// PowerStateGet gets the power state of a BMC machine.
+func (p *Provider) PowerStateGet(ctx context.Context) (state string, err error) {
+	rp := RequestPayload{
+		ID:     time.Now().UnixNano(),
+		Host:   p.Host,
+		Method: PowerGetMethod,
+	}
+	resp, err := p.process(ctx, rp)
+	if err != nil {
+		return "", err
+	}
+	if resp.Error != nil && resp.Error.Code != 0 {
+		return "", fmt.Errorf("error from rpc consumer: %v", resp.Error)
+	}
+
+	return resp.Result.(string), nil
+}
+
+// process is the main function for the roundtrip of rpc calls to the ConsumerURL.
+func (p *Provider) process(ctx context.Context, rp RequestPayload) (ResponsePayload, error) {
+	// 1. create the HTTP request.
+	// 2. create the signature payload.
+	// 3. sign the signature payload.
+	// 4. add signatures to the request as headers.
+	// 5. request/response round trip.
+	// 6. handle the response.
+	req, err := p.createRequest(ctx, rp)
+	if err != nil {
+		return ResponsePayload{}, err
+	}
+
+	// create the signature payload
+	reqBuf := new(bytes.Buffer)
+	if _, err := io.Copy(reqBuf, req.Body); err != nil {
+		return ResponsePayload{}, fmt.Errorf("failed to read request body: %w", err)
+	}
+	headersForSig := http.Header{}
+	for _, h := range p.Opts.Signature.IncludedPayloadHeaders {
+		if val := req.Header.Get(h); val != "" {
+			headersForSig.Add(h, val)
+		}
+	}
+	sigPay := createSignaturePayload(reqBuf.Bytes(), headersForSig)
+
+	// sign the signature payload
+	sigs, err := sign(sigPay, p.Opts.HMAC.Hashes, p.Opts.HMAC.PrefixSigDisabled)
+	if err != nil {
+		return ResponsePayload{}, err
+	}
+
+	// add signatures to the request as headers.
+	for algo, keys := range sigs {
+		if len(sigs) > 0 {
+			h := p.Opts.Signature.HeaderName
+			if !p.Opts.Signature.AppendAlgoToHeaderDisabled {
+				h = fmt.Sprintf("%s-%s", h, algo.ToShort())
+			}
+			req.Header.Add(h, strings.Join(keys, ","))
+		}
+	}
+
+	// request/response round trip.
+	kvs := requestKVS(req)
+	kvs = append(kvs, []interface{}{"host", p.Host, "method", rp.Method, "consumerURL", p.ConsumerURL}...)
+	if rp.Params != nil {
+		kvs = append(kvs, []interface{}{"params", rp.Params}...)
+	}
+
+	resp, err := p.Client.Do(req)
```

Comment:
Do we specify a timeout on the request?

Reply:
We do. We use `http.NewRequestWithContext` to create the `req`.

Reply:
Yea OK. I was thinking about affording them some protection from malicious/buggy endpoints but that's possible of any provider.